source/slang/slang-emit-metal.cpp

```diff
@@ -741,84 +744,24 @@ void MetalSourceEmitter::_emitType(IRType* type, DeclaratorInfo* declarator)
-    {
-        m_writer->emit(" [[vertex_id]]");
-    }
-    else if (semanticName.caseInsensitiveEquals(toSlice("SV_INSTANCEID")))
-    {
-        m_writer->emit(" [[instance_id]]");
-    }
-    else if (semanticName.caseInsensitiveEquals(toSlice("SV_Target")))
-    {
-        m_writer->emit(" [[color(");
-        m_writer->emit(semanticIndex);
-        m_writer->emit(")]]");
-    }
-    else if (semanticName.caseInsensitiveEquals(toSlice("SV_PRIMITIVEID")))
+    if (auto sysSemanticDecor = inst->findDecoration<IRTargetSystemValueDecoration>())
     {
-        m_writer->emit(" [[primitive_id]]");
-    }
-    else if (semanticName.caseInsensitiveEquals(toSlice("SV_GROUPID")))
-    {
-        // TODO: not supported by metal.
-        // We need to implement the transformation logic in slang-ir-metal-legalize.cpp
-        // to convert SV_GroupID to something like METAL_threadgroup_position_in_grid.
-    }
-    else if (semanticName.caseInsensitiveEquals(toSlice("SV_GROUPINDEX")))
-    {
-        // TODO: not supported by metal.
-    }
-    else if (semanticName.caseInsensitiveEquals(toSlice("SV_DISPATCHTHREADID")))
-    {
-        m_writer->emit(" [[thread_position_in_grid]]");
-    }
-    else if (semanticName.caseInsensitiveEquals(toSlice("SV_GROUPTHREADID")))
-    {
-        m_writer->emit(" [[thread_position_in_threadgroup]]");
-    }
-    else if (semanticName.caseInsensitiveEquals(toSlice("SV_CLIPDISTANCE")))
-    {
-        m_writer->emit(" [[clip_distance]]");
-    }
-    else if (semanticName.caseInsensitiveEquals(toSlice("SV_RENDERTARGETARRAYINDEX")))
-    {
-        m_writer->emit(" [[render_target_array_index]]");
-    }
-    else if (semanticName.caseInsensitiveEquals(toSlice("SV_VIEWPORTARRAYINDEX")))
-    {
-        m_writer->emit(" [[viewport_array_index]]");
-    }
-    else if (semanticName.caseInsensitiveEquals(toSlice("SV_Depth")))
-    {
-        m_writer->emit(" [[depth(any)]]");
-    }
-    else if (semanticName.caseInsensitiveEquals(toSlice("SV_DepthGreaterEqual")))
-    {
-        m_writer->emit(" [[depth(greater)]]");
-    }
-    else if (semanticName.caseInsensitiveEquals(toSlice("SV_DepthLessEqual")))
-    {
-        m_writer->emit(" [[depth(less)]]");
-    }
-    else if (semanticName.caseInsensitiveEquals(toSlice("SV_Coverage")))
-    {
-        m_writer->emit(" [[sample_mask]]");
-    }
-    else if (semanticName.caseInsensitiveEquals(toSlice("SV_StencilRef")))
-    {
-        m_writer->emit(" [[stencil]]");
+        m_writer->emit(" [[");
+        m_writer->emit(sysSemanticDecor->getSemantic());
+        m_writer->emit("]]");
+        return true;
     }
-    else
+    return false;
+}
+
+void MetalSourceEmitter::_emitUserSemantic(UnownedStringSlice semanticName, IRIntegerValue semanticIndex)
+{
+    if (!semanticName.startsWithCaseInsensitive(toSlice("SV_")))
     {
         m_writer->emit(" [[user(");
-        m_writer->emit(semanticName);
+        m_writer->emit(String(semanticName).toUpper());
```

Comment:
Is there a reason why to use upper cases?
Can we just use it as it is?

Reply:
HLSL doesn't distinguish upper/lower casing for semantic names, but metal does. So the only way to ensure we generate valid code is to canonicalize the name.

Reply:
Oh. I see.
In that case, should we add "_0" kind of suffix to avoid conflicts?
Both "sem" and "Sem" will become "SEM", as an example.

Reply:
I take this back. I misunderstood what you meant.
Since Slang is case insensitive, it should be fine to make them all upper case.